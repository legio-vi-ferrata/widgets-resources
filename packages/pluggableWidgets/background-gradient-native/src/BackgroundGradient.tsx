import { ReactElement, createElement } from "react";
import { Pressable, StyleSheet } from "react-native";
import LinearGradient from "react-native-linear-gradient";

import { executeAction } from "@mendix/piw-utils-internal";
import { mergeNativeStyles } from "@mendix/pluggable-widgets-tools";

import defaultStyle, { CustomStyle } from "./ui/Styles";

import { BackgroundGradientProps } from "../typings/BackgroundGradientProps";

export type props = BackgroundGradientProps<CustomStyle>;

export function BackgroundGradient({ name, colorList, angle, content, opacity, onClick, style }: props): ReactElement {
    const colors = colorList.map(colorsObject => colorsObject.color);
    const offsets = colorList.map(colorsObject => Number(colorsObject.offset));
    const styles = mergeNativeStyles(defaultStyle, style);
    return (
        <Pressable
            onPress={() => {
                executeAction(onClick);
            }}
            testID={`background-gradient-${name}`}
            style={({ pressed }) =>
                StyleSheet.flatten([
                    styles.container,
                    { opacity: onClick?.canExecute && pressed ? 0.5 : Number(opacity) }
                ])
            }
        >
            <LinearGradient colors={colors} locations={offsets} useAngle={!!angle} angle={angle}>
                {content}
            </LinearGradient>
        </Pressable>
    );
}
