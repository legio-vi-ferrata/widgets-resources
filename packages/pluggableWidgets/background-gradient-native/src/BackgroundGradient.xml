<?xml version="1.0" encoding="utf-8" ?>
<widget id="com.mendix.widget.native.backgroundgradient.BackgroundGradient" pluginWidget="true" needsEntityContext="true" offlineCapable="true" supportedPlatform="Native" xmlns="http://www.mendix.com/widget/1.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.mendix.com/widget/1.0/ ../node_modules/mendix/custom_widget.xsd">
    <name>Background Gradient</name>
    <description />
    <studioProCategory>Structure</studioProCategory>
    <studioCategory>Structure</studioCategory>
    <icon />
    <properties>
        <propertyGroup caption="General">
            <propertyGroup caption="General">
                <property key="colorList" type="object" isList="true" required="true">
                    <caption>Color List</caption>
                    <description />
                    <properties>
                        <propertyGroup caption="Colors Item">
                            <property key="color" type="string">
                                <caption>Color</caption>
                                <description>Valid color value format: #d0d0d0, rgb(115,155,155), or rgba(195,226,226,0.5)</description>
                            </property>
                            <property key="offset" type="decimal" defaultValue="0.0">
                                <caption>Offseet</caption>
                                <description>Where the gradient ends - Should be between 0 and 1</description>
                            </property>
                        </propertyGroup>
                    </properties>
                </property>
                <property key="angle" type="integer" defaultValue="0">
                    <caption>Angle</caption>
                    <description>For angled gradient effect</description>
                </property>
                <property key="opacity" type="decimal" defaultValue="1.0">
                    <caption>Opacity</caption>
                    <description>Should be between 0 and 1</description>
                </property>
                <property key="content" type="widgets">
                    <caption>Content</caption>
                    <description />
                </property>
            </propertyGroup>
            <propertyGroup caption="Visibility">
                <systemProperty key="Visibility" />
            </propertyGroup>
            <propertyGroup caption="Common">
                <systemProperty key="Name" />
            </propertyGroup>
        </propertyGroup>
        <propertyGroup caption="Events">
            <property key="onClick" type="action" required="false">
                <caption>On click</caption>
                <description />
            </property>
        </propertyGroup>
    </properties>
</widget>
