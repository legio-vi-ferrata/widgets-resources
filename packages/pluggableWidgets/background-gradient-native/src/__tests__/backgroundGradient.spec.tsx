import { createElement } from "react";
import { Text } from "react-native";
import { render, fireEvent } from "@testing-library/react-native";
import Big from "big.js";

import { ActionValue } from "mendix";

import { BackgroundGradient, props } from "../BackgroundGradient";

describe("Background gradient", () => {
    let defaultProps: props;
    beforeEach(() => {
        defaultProps = {
            angle: 0,
            colorList: [
                { color: "#fff", offset: Big(0) },
                { color: "#000", offset: Big(1) }
            ],
            content: <Text>Test</Text>,
            name: "test",
            opacity: Big(1),
            style: []
        };
    });
    it("render background gradient correctly", () => {
        const component = render(<BackgroundGradient {...defaultProps} />);
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("call onClick when the container is pressed", () => {
        const actionExecution = jest.fn();
        const action: ActionValue = {
            canExecute: true,
            execute: actionExecution,
            isExecuting: false
        };
        const component = render(<BackgroundGradient {...defaultProps} onClick={action} />);
        const pressable = component.getByTestId("background-gradient-test");
        fireEvent.press(pressable);
        expect(actionExecution).toBeCalled();
    });
});
