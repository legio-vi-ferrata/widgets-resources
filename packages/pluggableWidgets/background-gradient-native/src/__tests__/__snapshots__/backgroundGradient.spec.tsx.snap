// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Background gradient render background gradient correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
  testID="background-gradient-test"
>
  <BVLinearGradient
    angle={0}
    colors={
      Array [
        4294967295,
        4278190080,
      ]
    }
    endPoint={
      Object {
        "x": 0.5,
        "y": 1,
      }
    }
    locations={
      Array [
        0,
        1,
      ]
    }
    startPoint={
      Object {
        "x": 0.5,
        "y": 0,
      }
    }
    useAngle={false}
  >
    <Text>
      Test
    </Text>
  </BVLinearGradient>
</View>
`;
